src/relay/blockchain/events_informations.py

```diff
@@ -0,0 +1,133 @@
+import math
+from collections import namedtuple
+from typing import List
+
+from relay.blockchain.currency_network_events import (
+    BalanceUpdateEvent,
+    BalanceUpdateEventType,
+    TrustlineUpdateEventType,
+)
+from relay.network_graph.interests import calculate_interests
+
+InterestAccrued = namedtuple("InterestAccrued", ["value", "interest_rate", "timestamp"])
+
+DIRECTION_SENT = "sent"
+DIRECTION_RECEIVED = "received"
+
+
+def get_list_of_paid_interests_for_trustline(
+    events_proxy, currency_network_address, user, counterparty
+) -> List[InterestAccrued]:
+    """Get all balance changes of a trustline because of interests and the time at which it occurred.
+    Assumes the trustline exists"""
+
+    balance_update_events = events_proxy.get_trustline_events(
+        currency_network_address, user, counterparty, event_name=BalanceUpdateEventType
+    )
+
+    timestamps = [
+        balance_update_event.timestamp for balance_update_event in balance_update_events
+    ]
+
+    balances = get_sorted_balances_from_update_events(balance_update_events)
+    interest_rates = [
+        get_interests_rates_of_trustline_for_user_before_timestamp(
+            events_proxy,
+            currency_network_address,
+            user,
+            counterparty,
+            balance,
+            timestamp,
+        )
+        for (balance, timestamp) in zip(balances[:-1], timestamps[1:])
+    ]
+
+    return [
+        InterestAccrued(
+            calculate_interests(balance, interest_rates, post_time - pre_time),
+            interest_rates,
+            post_time,
+        )
+        for (balance, interest_rates, pre_time, post_time) in zip(
+            balances[:-1], interest_rates, timestamps[:-1], timestamps[1:]
+        )
+    ]
+
```

Comment:
This function is written using multiple list comprehensions, which makes it hard to understand.
From what I can tell you could write this as a single for loop:
```
for (pre_balance_event, post_balance_event) in zip(balance_update_events[:-1], balance_update_events[1:]):
   balance = balance_viewed_from_user(...)
   interest = ...
   retval.append(InterestAccrued(...))
```

Reply:
I changed it for a for loop as you suggest.

Reply:
no, you do not iterate over consecutive pairs of BalanceUpdateEvents as I suggested. If you do that, you can easily get rid of the assignments to balances and timestamps. 

Please use 
```
for (pre_balance_event, post_balance_event) in zip(balance_update_events, balance_update_events[1:]):
  ...
```
(i.e. the zip expression can also be simplified. It's using only two arguments and only one slice operation)

Reply:
Strictly speaking you could even introduce a function to iterate over consecutive pairs. But, we have toolz as an indirect dependency, so please use sliding window: https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.sliding_window